### DISCUSSION TITLE
alternative api to avoid versionstamp deadlock

### DISCUSSION THREAD

**Discussion by Pyrolistical** [josephg/node-foundationdb#62] [Ideas]
This is the current versionstamp API, which can lead to deadlock if one awaits the `tn.getVersionstamp().promise` inside `doTn`
```js
const p = await db.doTn(async tn => {
  tn.set('key', 'val')
  return tn.getVersionstamp()
})

const versionstamp = await p.promise // 10 byte buffer
```

Could we avoid this problem all together by another api like `doTnWithVersionstamp`
```js
const [value, versionstamp] = await db.doTnWithVersionstamp(async (tn, versionstamp) => {
  // versionstamp is just a placeholder, notice how much easier it is to use in a tuple key
  tn.set(['key', versionstamp, 'suffix'], 'val')
  return tn.get(['key', versionstamp, 'suffix'])
})

// value === 'val'
// versionstamp === 10 byte buffer
```

In this new api `await versionstamp` does nothing inside `doTnWithVersionstamp`

**Comment by josephg** [comment]
Out of curiosity, is this a problem you're running into or just a neat idea?

**Reply by Pyrolistical** [reply]
Just a neat idea. I'm an API guy so I saw this opportunity.
